int64 dx
int64 dy